### PR TITLE
Add FLIC Dataset

### PR DESCRIPTION
**Pull request by WilliamHYZhang**
Hi,

I'm currently working on adding the FLIC Dataset (https://bensapp.github.io/flic-dataset.html).

Currently, the plan is to have the annotations as closely resemble the original examples provided in the examples.mat file as we can.

Here is a basic outline of what I did already:
- citations, description, urls, all loaded
- example.mat file is loaded with scipy.io (will change import to tfds.core.lazy_imports in a future commit) (note: example.mat file is analyzed as a multidimensional array, e.g. data["example"][index_of_example_image][feature], where feature corresponds with the values in the FeatureDict)
- established FeatureDict in _info with features corresponding to the values given by .dtype (attribute names, mostly either tdfs.features.Image(), tfds.features.Text(), or tfds.features.Tensor() with corresponding shapes, types such as tf.uint16 and tf.float64 have also been found and implemented)
- established train/test splits in _split_generators and passed parameters to _generate_examples, done in the form of two lists (test and train) containing indexes of examples in data (note: no validation split from what I observed, booleans for istest/istrain are never both true or false)
- have a skeletal _generate_example function that simply yield the index of the attributes in data for each example in split

Some issues (most important to least important)
- download_and_prepare is failing as attributes with type tfds.features.Sequence are unable be yielded in the _generate_example function, returns error that the shapes must have the same length (note: these attributes are "poselet_hit_idx" and "coords", both of which are np arrays that have variable length for each example, see lines 99 and 101.)
- the "idx" values in the "poselet_hit_idx" array are nowhere in the range [1, 29] as the name-key mapping might suggest, I have absolutely no clue whatsoever what these numbers are supposed to represent
- "torsobox" attribute is in form of of 4 float Tensor instead of BBox, don't know if should change, if yes, then also need to implement normalize_bbox function and include PIL library (also probably from tfds.core.lazy_imports) to determine image sizes

Thanks, look forward to continuing with this project and contributing to Tensorflow.

### CODE REVIEW

File: tensorflow_datasets/image/flic.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    ]
+
+  def _generate_examples(self, extract_path, data, split):
+    """Yields examples."""
+    for index in split:
+      yield index, {
+        "image": os.path.join(extract_path, "FLIC", "images", data["examples"][index][3]),
+        "poselet_hit_idx": {"idx": list(data["examples"][index][0])},
+        "moviename": data["examples"][index][1],
+        "coords": {"coord": list(data["examples"][index][2])},
```
Comment (WilliamHYZhang) [inline tensorflow_datasets/image/flic.py]: This line should be rewritten to have "coords" return a dict of "x" and "y" tf.float64's. How should i use the zip() function for this?

**Comment by WilliamHYZhang** [comment]
update: pushed with the fix, see output .json file [here](https://gist.github.com/WilliamHYZhang/a91ba59e867acfc00b8d773dae257938)

File: tensorflow_datasets/image/flic.py (hunk 1 above)
Comment (WilliamHYZhang) [reply tensorflow_datasets/image/flic.py]: Disregard, fixed with push

**Comment by captain-pool** [comment]
Oops, I actually forgot to click Start Review Button. You should have got this review much earlier.😅

**Comment by WilliamHYZhang** [comment]
@captain-pool Haha, thanks for the feedback, already fixed the issue with the coords, working on the updated split.

File: tensorflow_datasets/image/flic.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                                       mat_dtype=True)
+
+    train_list = []
+    test_list = []
+
+    for i in range(len(data["examples"])):
+      if data["examples"][i][7]:
+        train_list.append(i)
+      elif data["examples"][i][8]:
+        test_list.append(i)
```
[the remark this answers is not in the document]
Comment (WilliamHYZhang) [reply tensorflow_datasets/image/flic.py]: @captain-pool thanks for the suggestion, pushed update with fix.

**Comment by captain-pool** [comment]
Issue reference: #547

**Comment by WilliamHYZhang** [comment]
@captain-pool thank you for the reviews! Will work on implementing changes and then testing later today.

File: tensorflow_datasets/image/flic.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+      tfds.core.SplitGenerator(
+        name=tfds.Split.TEST,
+        gen_kwargs={
+          "extract_path": extract_path,
+          "data": data,
+          "istrain": False,
+          "istest": True,
+        },
+      ),
+    ]
```
[the remark this answers is not in the document]
Comment (WilliamHYZhang) [reply tensorflow_datasets/image/flic.py]: I originally had a split for validation, however I believe in the annotations that none of the examples have both is true false and is test false, one of them is always true. This leads me to think that there was only intended to be too splits for the data.

File: tensorflow_datasets/image/flic.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  """Frames Labeled In Cinema (FLIC)"""
+
+  VERSION = tfds.core.Version('1.0.0')
+
+  def _info(self):
+    return tfds.core.DatasetInfo(
+        builder=self,
+        description=_DESCRIPTION,
+        features=tfds.features.FeaturesDict({
+            "image": tfds.features.Image(),
```
Comment (captain-pool) [inline tensorflow_datasets/image/flic.py]: As far as I remember the shape of all the images are more or less same. So, you can put that shape here as:
`tfds.features.Image(shape=(height, width, 3))`

File: tensorflow_datasets/image/flic.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (captain-pool) [reply tensorflow_datasets/image/flic.py]: > 
> 
> I originally had a split for validation, however I believe in the annotations that none of the examples have both is true false and is test false, one of them is always true. This leads me to think that there was only intended to be too splits for the data.

I see.

**Comment by WilliamHYZhang** [comment]
@captain-pool pushed new update with updated main file, new testing and fake examples also added. All tests are passing. Thanks so much!

**Comment by captain-pool** [comment]
@cyfra, @rsepassi please take a look

File: tensorflow_datasets/image/flic.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+detector on every tenth frame of 30 movies. People detected with high confidence
+(roughly 20K candidates) were then sent to the crowdsourcing marketplace Amazon
+Mechanical Turk to obtain groundtruthlabeling. Each image was annotated by five
+Turkers for $0.01 each to label 10 upperbody joints. The median-of-five labeling
+was taken in each image to be robust to outlier annotation. Finally, images were
+rejected manually by us if the person was occluded or severely non-frontal. We
+set aside 20% (1016 images) of the data for testing. 
+"""
+
+_URL = "https://drive.google.com/uc?id=0B4K3PZp8xXDJN0Fpb0piVjQ3Y3M&export=download"
```
Comment (captain-pool) [inline tensorflow_datasets/image/flic.py]: Hey, I just suddenly remembered there are two versions of this right? One full dataset and another partial? And this is the partial right?

Will it be possible for you to write a heavy configuration to provide both versions of the datasets?
For example: "flic/full" and "flic/small".

Comment (WilliamHYZhang) [reply tensorflow_datasets/image/flic.py]: Of course! On it.

**Comment by WilliamHYZhang** [comment]
@captain-pool All done! Implemented the builderconfig and modified the testing and fake data for the new structure. All tests passing.

**Comment by WilliamHYZhang** [comment]
Is there anything I should do before we can merge this and my other approved PRs in this repo? Thanks!

File: tensorflow_datasets/image/flic.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+  """BuilderConfig for FLIC."""
+
+  @tfds.core.disallow_positional_args
+
+  def __init__(self, data, **kwargs):
+    """Constructs a FlicConfig."""
+    if data not in _DATA_OPTIONS:
+      raise ValueError("data must be one of %s" % _DATA_OPTIONS)
+
+    name = kwargs.get("name", data)
```
Comment (cyfra) [inline tensorflow_datasets/image/flic.py]: nit: I'd put name=data in line 63 - and remove lines 48&49.

File: tensorflow_datasets/image/flic.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    configs.append(FlicConfig(
+        version=tfds.core.Version("2.0.0"),
+        data=data))
+  return configs
+
+class Flic(tfds.core.GeneratorBasedBuilder):
+  """Frames Labeled In Cinema (FLIC)"""
+
+  BUILDER_CONFIGS = _make_builder_configs()
+  VERSION = tfds.core.Version("2.0.0")
```
Comment (cyfra) [inline tensorflow_datasets/image/flic.py]: can be removed (as you have versions in your configs)

File: tensorflow_datasets/image/flic.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        }),
+        citation=_CITATION,
+    )
+
+  def _split_generators(self, dl_manager):
+    """Returns SplitGenerators."""
+    extract_path = dl_manager.download_and_extract(
+        self.builder_config.url)
+    
+    data = scipy.io.loadmat(os.path.join(extract_path,
```
Comment (cyfra) [inline tensorflow_datasets/image/flic.py]: please use tfds.core.lazy_imports.scipy.io.loadmat and remove the direct import of scipy.io

File: tensorflow_datasets/image/flic.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+                                         self.builder_config.dir, "examples.mat"),
```
Comment (cyfra) [inline tensorflow_datasets/image/flic.py]: Please change the code to use tf.io.gfile to read the data (this makes the code work on non-local filesystems too):

    with tf.io.gfile.GFile(filepath, "rb") as f:
      data = tfds.core.lazy_imports.scipy.io.loadmat(f)

File: tensorflow_datasets/image/flic.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+                            struct_as_record=True, squeeze_me=True, mat_dtype=True)
+
+    return [
+      tfds.core.SplitGenerator(
+        name=tfds.Split.TRAIN,
+        gen_kwargs={
+          "extract_path": extract_path,
+          "data": data,
+          "istrain": True,
```
Comment (cyfra) [inline tensorflow_datasets/image/flic.py]: nit:
having two separate bools, that are never both 'true' or both 'false' is not a great design.
Consider passing an enum or simply the 'name' argument instead.

File: tensorflow_datasets/image/flic.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+          "istrain": False,
+          "istest": True,
+        },
+      ),
+    ]
+
+  def _generate_examples(self, extract_path, data, istrain, istest):
+    """Yields examples."""
+    for u_id, example in enumerate(data["examples"]):
+      if (example[7] and istrain) or (example[8] and istest):
```
Comment (cyfra) [inline tensorflow_datasets/image/flic.py]: avoid 'magic numbers' - what is 7? what is 8?

File: tensorflow_datasets/image/flic.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+Mechanical Turk to obtain groundtruthlabeling. Each image was annotated by five
+Turkers for $0.01 each to label 10 upperbody joints. The median-of-five labeling
+was taken in each image to be robust to outlier annotation. Finally, images were
+rejected manually by us if the person was occluded or severely non-frontal. We
+set aside 20% (1016 images) of the data for testing. 
+"""
+
+_DATA_OPTIONS = ["small", "full"]
+
+_URL_SUBSET = "https://drive.google.com/uc?id=0B4K3PZp8xXDJN0Fpb0piVjQ3Y3M&export=download"
```
[the remark this answers is not in the document]
Comment (WilliamHYZhang) [reply tensorflow_datasets/image/flic.py]: They are the official links the dataset has provided.

**Comment by WilliamHYZhang** [comment]
>nit:
having two separate bools, that are never both 'true' or both 'false' is not a great design.
Consider passing an enum or simply the 'name' argument instead.

>avoid 'magic numbers' - what is 7? what is 8?

I agree with this, however the dataset itself is structured like this, where for each example the 7th and 8th index respectively are booleans, `istrain` and `istest`. I was trying to reflect this in the splits.

**Comment by WilliamHYZhang** [comment]
Thank you @cyfra for the time taken to review and your comments, appreciated very much.

**Comment by cyfra** [comment]
> I agree with this, however the dataset itself is structured like this, where for each example the 7th and 8th index respectively are booleans, `istrain` and `istest`. I was trying to reflect this in the splits.


What you could do: is pass the 'int' argument (for example 'selection_column').
And then in the code you do:

if example[selection_column]:
  ....

**Comment by WilliamHYZhang** [comment]
@ updated to reflect your suggestion.

File: tensorflow_datasets/image/flic.py [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    )
+
+  def _split_generators(self, dl_manager):
+    """Returns SplitGenerators."""
+    extract_path = dl_manager.download_and_extract(
+        self.builder_config.url)
+    
+    mat_path = os.path.join(extract_path, self.builder_config.dir, "examples.mat")
+    with tf.io.gfile.GFile(mat_path) as f:
+    data = tfds.core.lazy_imports.scipy.io.loadmat(f, struct_as_record=True,
```
[the remark this answers is not in the document]
Comment (WilliamHYZhang) [reply tensorflow_datasets/image/flic.py]: Sorry, didn't catch that, should pass now.

File: tensorflow_datasets/image/flic.py [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+          "data": data,
+          "selection_column": 7,
+        },
+      ),
+      tfds.core.SplitGenerator(
+        name=tfds.Split.TEST,
+        gen_kwargs={
+          "extract_path": extract_path,
+          "data": data,
+          "selection_column": 8,
```
Comment (Conchylicultor) [inline tensorflow_datasets/image/flic.py]: Could you add a comment about what this selection_column represent ?

**Comment by WilliamHYZhang** [comment]
Sure, no problem.

**Comment by Conchylicultor** [comment]
Thanks, almost good to go. 

I think for consistency with other datasets, torsobox should be a `BBoxFeature()` instead of a `Tensor(shape=(4,))`, otherwise it's not clear for the user what are the coordinate order `(x, y, xmin, ymin)` ? `(x, xmin, y, ymin)`. What do you think ?

**Comment by WilliamHYZhang** [comment]
Oh yeah, good point. Thanks for the suggestion.

**Comment by WilliamHYZhang** [comment]
@Conchylicultor Implemented BBox attribute, fixed some bugs, uploaded checksums.

**Pull request merged**